subarray.go

```diff
@@ -0,0 +1,255 @@
+// manually release memory if needed. Free is idempotent and can safely be called many times on the same object.
+func (sa *Subarray) Free() {
+	if sa.subarray != nil {
+		C.tiledb_subarray_free(&sa.subarray)
+	}
+}
+
+// SetCoalesceRanges sets coalesce_ranges property on a TileDB subarray object.
+// Intended to be used just after array.NewSubarray to replace the initial coalesce_ranges == true with coalesce_ranges = false if needed.
+func (sa *Subarray) SetCoalesceRanges(b bool) error {
+	var coalesce C.int
+	if b {
+		coalesce = 1
+	}
+
+	ret := C.tiledb_subarray_set_coalesce_ranges(sa.context.tiledbContext, sa.subarray, coalesce)
+	if ret != C.TILEDB_OK {
+		return fmt.Errorf("Error setting coalesce ranges on subarray: %s", sa.context.LastError())
+	}
+
+	return nil
+}
+
+// AddRange adds a range along a subarray dimension. It checks the types of range and dimension and
+// if the datatype of the range is not the same as the type of the dimension it returns an error.
+func (sa *Subarray) AddRange(dimIdx uint32, r Range) error {
+	dt, isVar, err := datatypeOfDimensionFromIndex(sa.array, dimIdx)
+	if err != nil {
+		return err
+	}
+	if err := r.assertCompatibility(dt, isVar); err != nil {
+		return err
+	}
+
+	var ret C.int32_t
+	if isVar {
+		startSlice := []byte(r.start.(string))
+		endSlice := []byte(r.end.(string))
+		ret = C.tiledb_subarray_add_range_var(sa.context.tiledbContext, sa.subarray, C.uint32_t(dimIdx),
+			unsafe.Pointer(&startSlice[0]), C.uint64_t(len(startSlice)), unsafe.Pointer(&endSlice[0]), C.uint64_t(len(endSlice)))
+	} else {
+		ret = C.tiledb_subarray_add_range(sa.context.tiledbContext, sa.subarray, C.uint32_t(dimIdx),
+			addressableValue(r.start).UnsafePointer(), addressableValue(r.end).UnsafePointer(), nil)
+	}
+	if ret != C.TILEDB_OK {
+		return fmt.Errorf("Error adding subarray range: %s", sa.context.LastError())
+	}
+
+	return nil
+}
+
+// AddRangeByName adds a range along a subarray dimension. It checks the types of range and dimension and
+// if the datatype of the range is not the same as the type of the dimension it returns an error.
+func (sa *Subarray) AddRangeByName(dimName string, r Range) error {
+	dt, isVar, err := datatypeOfDimensionFromName(sa.array, dimName)
+	if err != nil {
+		return err
+	}
+	if err := r.assertCompatibility(dt, isVar); err != nil {
+		return err
+	}
+
+	cDimName := C.CString(dimName)
+	defer C.free(unsafe.Pointer(cDimName))
+
+	var ret C.int32_t
+	if isVar {
+		startSlice := []byte(r.start.(string))
+		endSlice := []byte(r.end.(string))
+		ret = C.tiledb_subarray_add_range_var_by_name(sa.context.tiledbContext, sa.subarray, cDimName,
+			unsafe.Pointer(&startSlice[0]), C.uint64_t(len(startSlice)), unsafe.Pointer(&endSlice[0]), C.uint64_t(len(endSlice)))
+	} else {
+		ret = C.tiledb_subarray_add_range_by_name(sa.context.tiledbContext, sa.subarray, cDimName,
+			addressableValue(r.start).UnsafePointer(), addressableValue(r.end).UnsafePointer(), nil)
```

Comment:
Added defensive keepalives. I think it is sufficient to keep alive `r`

Reply:
But the new addressableValue values are completely independent of R; they are copies of the data from R. That is, we’re doing something vaguely like

```go
startPtr := new(typeof r.start)
*startPtr = r.start
unsafeStart := unsafe.Pointer(startPtr)
// from here on out, we no longer refer to `startPtr` so it is eligible for GC.
```
making startPtr a new pointer on the heap. If r.start were a pointer itself, the data pointed to by r would be preserved, but not necessarily the `startPtr` pointer to that same data. However, the data will *also* be preserved by keeping alive startPtr, without separately needing to keep r alive.

Reply:
Made the change for extra safety. But i think i was OK. `r.start` is passed to `addressableValue` and the `reflect` package makes copies. This is in the call place
```
                ret = C.tiledb_subarray_add_range(sa.context.tiledbContext, sa.subarray, C.uint32_t(dimIdx),
                        addressableValue(r.start).UnsafePointer(), addressableValue(r.end).UnsafePointer(), nil)
```
so the Keepalive just after maybe is reduntant

Reply:
The thing that needs to be kept alive isn't `r.start` and `r.end`; it’s the `addressableValue`s, because they are otherwise immediately forgotten about. Even though it's on the same line, what this looks like internally is

```go
startPtr := addressableValue(r.start)
startUnsafe := startPtr.UnsafePointer()
// Since startUnsafe is unmanaged, as of here, `startPtr` and its associated Value
// are never referred to again and thus eligible for collection.
ret = C.tiledb_whatever(..., startUnsafe, ...)
// This means that it is possible that this C call read or write to/from newly freed memory.
```

I admit it is *unlikely* that a GC cycle runs right there, but technically possible.